File: tools/viz_check/ikc/kernel_check.mjs
/* IKC kernel cross-check, JS replay leg.
 *
 * Reads <pyJson> (output of instrumented/kernel_check.py) and runs
 * COMDET.IKC.runIKC against the same input. IKC is deterministic
 * (no RNG), so JS replay needs no oracle injection — just match
 * per-iteration (max_k, components, kept) and final partition.
 *
 * Acceptance:
 *   - per-iter max_k matches
 *   - per-iter component sizes match (sorted ASC)
 *   - per-iter kept-component sizes match (sorted ASC)
 *   - final partition CSV byte-equal vs canonical (when sorted +
 *     header-stripped + cluster-id renumbered)
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const args = process.argv.slice(2);
if (args.length < 3) {
  console.error("usage: kernel_check.mjs <pyJson> <edge.csv> <out.csv>");
  process.exit(2);
}
const [pyPath, edgePath, outCsv] = args;

const py = JSON.parse(fs.readFileSync(pyPath, "utf8"));
const kFloor = py.k_floor;

globalThis.window = globalThis;
globalThis.window.COMDET = { FIXTURE: { nodes: [], edges: [], gt: [] } };
const WEB = path.join(__dirname, "../../../vltanh.github.io/comdet/js");
await import(path.join(WEB, "ikc/ikc.js"));

// Load edges from CSV (positional first two columns; skip self-loops to
// match canonical run_ikc.format_graph's removeSelfLoops + JS runIKC's
// e[0]===e[1] guard).
function loadEdges(edgePath) {
  const lines = fs.readFileSync(edgePath, "utf8").trim().split(/\r?\n/);
  const nodeSet = new Set();
  const edges = [];
  for (let i = 1; i < lines.length; i++) {
    if (!lines[i]) continue;
    const cols = lines[i].split(/[,\t ]/);
    const u = parseInt(cols[0], 10);
    const v = parseInt(cols[1], 10);
    if (u === v) continue;
    nodeSet.add(u);
    nodeSet.add(v);
    edges.push([u, v]);
  }
  // Encounter order to match canonical's pd.unique.
  const seen = new Set();
  const nodes = [];
  for (let i = 1; i < lines.length; i++) {
    if (!lines[i]) continue;
    const cols = lines[i].split(/[,\t ]/);
    const u = parseInt(cols[0], 10);
    const v = parseInt(cols[1], 10);
    if (!seen.has(u)) { seen.add(u); nodes.push(u); }
    if (!seen.has(v)) { seen.add(v); nodes.push(v); }
  }
  return { nodes, edges };
}

const { nodes, edges } = loadEdges(edgePath);
const res = COMDET.IKC.runIKC(nodes, edges, { kFloor, canonicalGate: true });

// Compare per-iteration with python trace.
const pyIters = py.iters;
const jsIters = res.iterations;

let fail = false;
function fmt(arr) { return JSON.stringify(arr); }

// Python iter records have {iter, max_k, components:[{size,kept,k_valid,modular_pos}], kept_clusters_this_iter}
// or {iter, max_k, terminated} on bail. Trim JS to comparable shape.
const pyActive = pyIters.filter(p => !p.terminated);
const jsActive = jsIters.filter(j => !j.bailed);

if (pyActive.length !== jsActive.length) {
  console.error(`ITER_COUNT mismatch: py=${pyActive.length} js=${jsActive.length}`);
  fail = true;
}

const lim = Math.min(pyActive.length, jsActive.length);
for (let i = 0; i < lim; i++) {
  const p = pyActive[i];
  const j = jsActive[i];
  if (p.max_k !== j.maxK) {
    console.error(`iter=${i} max_k mismatch py=${p.max_k} js=${j.maxK}`);
    fail = true;
  }
  const pSizes = p.components.map(c => c.size).sort((a,b)=>a-b);
  const jSizes = j.components.map(c => c.nodes.length).sort((a,b)=>a-b);
  if (fmt(pSizes) !== fmt(jSizes)) {
    console.error(`iter=${i} comp sizes py=${fmt(pSizes)} js=${fmt(jSizes)}`);
    fail = true;
  }
  const pKept = p.components.filter(c => c.kept).map(c => c.size).sort((a,b)=>a-b);
  const jKept = j.components.filter(c => c.accepted).map(c => c.nodes.length).sort((a,b)=>a-b);
  if (fmt(pKept) !== fmt(jKept)) {
    console.error(`iter=${i} kept sizes py=${fmt(pKept)} js=${fmt(jKept)}`);
    fail = true;
  }
}

// Bail-iter alignment.
const pBail = pyIters.find(p => p.terminated);
const jBail = jsIters.find(j => j.bailed);
if (!!pBail !== !!jBail) {
  console.error(`bail mismatch: py=${!!pBail} js=${!!jBail}`);
  fail = true;
} else if (pBail && jBail) {
  if (pBail.max_k !== jBail.maxK) {
    console.error(`bail max_k mismatch: py=${pBail.max_k} js=${jBail.maxK}`);
    fail = true;
  }
}

// Write CSV in run_ikc.print_clusters format. Accepted-cluster ordering
// in JS comes from JS's connectedComponentsCompact iteration; canonical
// comes from NetworKit's WeaklyConnectedComponents.getComponents() over
// a re-compacted graph each iteration. Order can diverge while membership
// is identical. Bridge: match each tracer-canonical cluster (size > 1,
// in tracer order) to a JS accepted cluster by member-set equality, and
// emit CSV using that order. Verifies partition equality and byte-equal
// vs canonical CSV simultaneously.
const tracerOrder = py.final_clusters.filter(c => c.size > 1);
const jsKept = res.accepted.filter(cl => cl.members.length > 1);

function memberKey(arr) { return JSON.stringify(arr.slice().sort((a,b)=>a-b)); }
const jsByKey = new Map();
jsKept.forEach(cl => jsByKey.set(memberKey(cl.members), cl));

if (tracerOrder.length !== jsKept.length) {
  console.error(`KEPT_COUNT mismatch tracer=${tracerOrder.length} js=${jsKept.length}`);
  fail = true;
}

const lines2 = ["node_id,cluster_id"];
let cid = 0;
for (const tc of tracerOrder) {
  // tracer cluster nodes are compact-internal IDs (post format_graph,
  // which is identity here since g1 was already 0..n-1). Map each via
  // the encounter-order node_map back to orig name.
  const origNodes = tc.nodes.map(ci => parseInt(py.node_map[ci].orig, 10));
  const key = memberKey(origNodes);
  const m = jsByKey.get(key);
  if (!m) {
    console.error(`tracer cluster ${cid} (orig=${origNodes.slice(0,5)}...) has no JS match`);
    fail = true;
    cid += 1;
    continue;
  }
  // Emit node-rows in tracer (canonical) order for byte-equal CSV.
  origNodes.forEach(node => lines2.push(`${node},${cid}`));
  cid += 1;
}
fs.writeFileSync(outCsv, lines2.join("\n") + "\n");

if (fail) {
  console.error("FAIL");
  process.exit(1);
} else {
  console.error(`PASS iter_compare ${pyActive.length} active iters`);
}

File: tools/viz_check/ikc/kernel_check.py
"""IKC kernel cross-check driver.

Three legs:
  - canonical: src/ikc/run_ikc.py (Python; networkit + pandas).
  - tracer: tools/viz_check/ikc/instrumented/kernel_check.py (Python
    fork of the canonical with [TRACE-IKC] log lines + JSON emit).
  - JS replay: kernel_check.mjs (loads comdet/js/ikc/ikc.js via Node).

IKC is deterministic (no RNG), so JS replay matches canonical via
per-iteration max_k + component-size + kept-component-size tuples
plus member-set bijection at the partition level. CSV byte-equal vs
canonical is achieved by emitting accepted clusters in tracer
(canonical) order.

Run:
    python tools/viz_check/ikc/kernel_check.py [--verbose]
"""
from __future__ import annotations

import argparse
import os
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE.parent / "_common"))
import driver as D

REPO = D.repo_root_from_here(HERE)
CANON = REPO / "src" / "ikc" / "run_ikc.py"
TRACER = HERE / "instrumented" / "kernel_check.py"
JS_REPLAY = HERE / "kernel_check.mjs"
PYCOMMON = REPO / "src" / "_common"

# Default k_floor per fixture. dnc uses 3 (matches examples/regen.sh
# ikc-3 row); fixture32 uses 2 (smaller graph; demo-friendly).
KFLOOR = {"fixture32": 2, "dnc": 3}

# Conda env name (canonical IKC needs networkit).
CONDA_ENV = os.environ.get("IKC_CONDA_ENV", "nwbench")
PY = ["conda", "run", "-n", CONDA_ENV, "python"]


def normalize_edge_for_canon(edge: Path, work: Path, name: str) -> Path:
    """Canonical run_ikc.py requires `source,target` headers. fixture32
    edge.csv uses `node_id1,node_id2`; rewrite to a sibling tmp file
    when needed. dnc_edge.csv already uses source,target."""
    head = edge.read_text().split("\n", 1)[0].strip()
    if head == "source,target":
        return edge
    out = work / f"{name}_st_edge.csv"
    body = edge.read_text().split("\n", 1)[1]
    out.write_text("source,target\n" + body)
    return out


def canonical_fn(edge: Path, com: Path, work: Path, name: str):
    edge_st = normalize_edge_for_canon(edge, work, name)
    out_dir = work / f"{name}_ikc_canon"
    out_dir.mkdir(exist_ok=True)
    env = os.environ.copy()
    env["PYTHONPATH"] = str(PYCOMMON)
    rc = subprocess.run(PY + [str(CANON), "-e", str(edge_st),
                              "-o", str(out_dir),
                              "-k", str(KFLOOR[name]), "-q"],
                        capture_output=True, text=True, env=env)
    if rc.returncode != 0:
        raise RuntimeError(f"canonical failed: {rc.stderr}")
    return (out_dir / "com.csv").read_text(), rc.stdout + rc.stderr


def tracer_fn(edge: Path, com: Path, work: Path, name: str):
    out = work / f"{name}_ikc_tracer.csv"
    json_path = work / f"{name}_ikc_tracer.json"
    log_path = work / f"{name}_ikc_tracer.log"
    rc = subprocess.run(PY + [str(TRACER), str(edge),
                              str(KFLOOR[name]), str(out)],
                        capture_output=True, text=True)
    if rc.returncode != 0:
        log_path.write_text(rc.stderr)
        raise RuntimeError(f"tracer failed: {rc.stderr}")
    json_path.write_text(rc.stdout)
    log_path.write_text(rc.stderr)
    return out.read_text(), json_path, rc.stderr


def replay_fn(json_path: Path, edge: Path, com: Path):
    out = edge.parent / (json_path.stem.replace("_ikc_tracer", "_ikc_js") + ".csv")
    rc = subprocess.run(["node", str(JS_REPLAY), str(json_path),
                         str(edge), str(out)],
                        capture_output=True, text=True)
    return rc.returncode == 0, rc.stdout + rc.stderr


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--verbose", action="store_true")
    args = ap.parse_args()

    if not CANON.exists():
        sys.exit(f"missing canonical: {CANON}")
    work = D.workdir(REPO)
    D.emit_fixture_inputs(REPO, work)

    failures = 0
    for name, edge, com in D.fixture_cases(work):
        failures += D.run_three_legs(
            name, edge, com, work,
            canonical_fn=canonical_fn, tracer_fn=tracer_fn, replay_fn=replay_fn,
            verbose=args.verbose,
        )

    print()
    if failures:
        print(f"OVERALL: FAIL ({failures} failures)")
        sys.exit(1)
    print("OVERALL: PASS")


if __name__ == "__main__":
    main()
